core/trino-main/src/main/java/io/trino/type/DoubleOperators.java

```diff
@@ -175,13 +179,35 @@ public static long castToReal(@SqlType(StandardTypes.DOUBLE) double value)
     @SqlType("varchar(x)")
     public static Slice castToVarchar(@LiteralParameter("x") long x, @SqlType(StandardTypes.DOUBLE) double value)
     {
-        String stringValue = String.valueOf(value);
+        String stringValue;
+
+        // handle positive and negative 0
+        if (value == 0e0) {
+            if (1e0 / value > 0) {
+                stringValue = "0E0";
+            }
+            else {
+                stringValue = "-0E0";
+            }
+        }
+        else if (Double.isInfinite(value)) {
+            if (value > 0) {
+                stringValue = "Infinity";
+            }
+            else {
+                stringValue = "-Infinity";
+            }
+        }
+        else {
+            stringValue = FORMAT.get().format(value);
+        }
+
         // String is all-ASCII, so String.length() here returns actual code points count
         if (stringValue.length() <= x) {
             return utf8Slice(stringValue);
         }
 
-        throw new TrinoException(INVALID_CAST_ARGUMENT, format("Value %s cannot be represented as varchar(%s)", value, x));
+        throw new TrinoException(INVALID_CAST_ARGUMENT, format("Value %s cannot be represented as varchar(%s)", stringValue, x));
```

Comment:
provide original `value` in the message, just in case we screw something up:

`"Value %s (%s) ...", value, stringValue`

Reply:
Applied to both double and real.